### ISSUE TITLE
Adding universal constants

### ISSUE THREAD

**Issue by cquiroz** [erikerlandson/coulomb#102]
Do you think coulomb should include definitions of constants like the ones described in
https://en.wikipedia.org/wiki/List_of_physical_constants

Or https://physics.nist.gov/cuu/pdf/wall_2018.pdf

**Comment by erikerlandson** [comment]
Overall, I like the idea of a `coulomb-physical-constants` package.  I think the main design-level question to answer is what value type the constants should use.  I ultimately gravitated to using Rational for internal manipulation of conversion coefficients, because it could represent both exact ratios and "floating point" values equally well.  

I think the same properties would make Rational a good candidate for physical constants. The trade-off is that any conversions to some other format such as Double would require actual run-time computation (the manipulation of unit coefficients internal to coulomb is all compile-time).

The three most plausible candidates for representing them in a package (imo) are:
- Double
- Rational
- BigDecimal

**Comment by erikerlandson** [comment]
actually, making these functions instead of values would make it easier to accommodate different representation types.

```scala
object constants {
  // define physical constants as derived units
  trait PhysicalConstantC
  implicit val definePhysicalConstantC =
    DerivedUnit[PhysicalConstantC, Meter %/ Second](coef = 299792458, name= "speed-of-light", abbv = "c")

  // returns speed of light using value type V
  def c[V](implicit ctv: ConvertableTo[V]): V =
    ctv.fromType[Rational](definePhysicalConstantC.coef).withUnit[PhysicalConstantC]
}
```

I think there are some improvements we could make here but this is an approach that can be made type flexible and efficient, keeping the conversions at compile time

**Comment by cquiroz** [comment]
That's brilliant

**Comment by erikerlandson** [comment]
This is getting close:

```scala
object constants {
  // define physical constants as derived units
  trait SpeedOfLight
  implicit val defineSpeedOfLight =
    DerivedUnit[SpeedOfLight, Meter %/ Second](coef = 299792458, name= "speed-of-light", abbv = "c")

  // returns speed of light using value type V
  def speedOfLight[V](implicit pcq: PhysicalConstantQuantity[V, SpeedOfLight]): Quantity[V, pcq.QU] = pcq.q

  trait PhysicalConstantQuantity[V, CU] {
    type QU
    val q: Quantity[V, QU]
  }
  object PhysicalConstantQuantity {
    implicit def summon[V, CU, U](implicit
      dcu: DerivedUnit[CU, U],
      ctv: ConvertableTo[V]): PhysicalConstantQuantity[V, CU] =
      new PhysicalConstantQuantity[V, CU] {
        type QU = U
        val q: Quantity[V, QU] = new Quantity[V, U](ctv.fromType[Rational](dcu.coef))
      }
  }
}
```

**Comment by cquiroz** [comment]
Should we use this approach to make a new constants module?

**Comment by erikerlandson** [comment]
I think this is a good approach.  For a final library I was thinking of an organization like

```scala
package object physicalconstants {
  // returns speed of light using value type V
  def speedOfLight[V](implicit pcq: PhysicalConstantQuantity[V, SpeedOfLight]): Quantity[V, pcq.QU] = pcq.q

  // actual unit defs in a subpackage
  package units {
    // define physical constants as derived units
    trait SpeedOfLight
    implicit val defineSpeedOfLight =
      DerivedUnit[SpeedOfLight, Meter %/ Second](coef = 299792458, name= "speed-of-light", abbv = "c")
  }

  package infra {
    trait PhysicalConstantQuantity[V, CU] {
      type QU
      val q: Quantity[V, QU]
    }
    object PhysicalConstantQuantity {
      implicit def summon[V, CU, U](implicit
        dcu: DerivedUnit[CU, U],
        ctv: ConvertableTo[V]): PhysicalConstantQuantity[V, CU] =
        new PhysicalConstantQuantity[V, CU] {
          type QU = U
          val q: Quantity[V, QU] = new Quantity[V, U](ctv.fromType[Rational](dcu.coef))
        }
    }
  }
}
```

**Comment by erikerlandson** [comment]
closed with #113

**Comment by erikerlandson** [comment]
included on coulomb release 0.5.5
